src/test/java/picard/sam/RevertSamTest.java

```diff
@@ -592,4 +604,135 @@ public void testHardClipRoundTrip() throws Exception {
+     */
+    private PicardHtsPath getTestReadGroupMapFile(final PicardHtsPath sam, final String outputBase) {
+        final String samExtension = sam.isBam() ? ".bam" : ".cram";
+        final PicardHtsPath result = PicardBucketUtils.getTempFilePath(DEFAULT_CLOUD_TEST_OUTPUT_DIR.getURIString(), ".txt");
+        final List<String> readGroups = getReadGroups(sam.toPath()).stream().map(rg -> rg.getId()).collect(Collectors.toList());
+        final List<String> readGroupOutput = new ArrayList<>();
+
+        for (String readGroup : readGroups){
+            final PicardHtsPath readGroupOutputPath = PicardBucketUtils.getTempFilePath(outputBase, readGroup, samExtension);
+            readGroupOutput.add(readGroupOutputPath.getURIString());
+        }
+
+        RevertSam.writeOutputMapToFile(result.toPath(), readGroups, readGroupOutput);
+        return result;
+    }
+
+    private List<SAMReadGroupRecord> getReadGroups(final Path sam){
+        final List<SAMReadGroupRecord> readGroupsInInput = SamReaderFactory.makeDefault().open(sam).getFileHeader().getReadGroups();
+        return readGroupsInInput;
+    }
+
+    /***
+     *
+     * @param inputSAM The path to the input SAM file
+     * @param outputPath The path to the single output SAM file, when output by read group is turned off. May be null.
+     * @param outputByReadGroup Whether to output by read group
+     * @param outputMapOptions Dictates whether to dynamically create an output map and where (local or cloud)
+     */
+    @Test(dataProvider = "cloudTestData", groups = "cloud")
+    public void testCloud(final PicardHtsPath inputSAM, final PicardHtsPath outputPath, final boolean outputByReadGroup,
+                          final OutputMapOptions outputMapOptions) {
+        final List<SAMReadGroupRecord> readGroupsInInput = SamReaderFactory.makeDefault().open(inputSAM.toPath()).getFileHeader().getReadGroups();
+        final List<String> args = new ArrayList<>(Arrays.asList(
+                "INPUT=" + inputSAM.getURIString(),
+                "OUTPUT_BY_READGROUP=" + outputByReadGroup));
+
+        if (outputPath != null){
+            args.add("OUTPUT=" + outputPath.getURIString());
+        }
+
+        final boolean createOutputMap = outputMapOptions != OutputMapOptions.NO_OUTPUT_MAP;
+        PicardHtsPath outputMap = null;
+        if (createOutputMap){
+            // If these conditions are met, we dynamically create the output map file based on the read groups in inputSAM
+            final String outputMapDir = outputMapOptions == OutputMapOptions.CLOUD_OUTPUT_MAP ?
+                    DEFAULT_CLOUD_TEST_OUTPUT_DIR.getURIString() :
+                    REVERT_SAM_LOCAL_TEST_DATA_DIR;
+            outputMap = getTestReadGroupMapFile(inputSAM, outputMapDir);
+            args.add("OUTPUT_MAP=" + outputMap.getURIString());
+        }
+
+        if (inputSAM.isCram()){
+            args.add("REFERENCE_SEQUENCE=" + HG19_CHR2021);
```

Comment:
The reference to use should really be part of the data provider; otherwise its hard to add new test files to this.

Reply:
done